src/api/classify_element.rs

```diff
@@ -0,0 +1,349 @@
+                    if let Err(err) = self.to_providers[port].try_send(Some(packet)) {
+                        panic!("Error in to_providers[{}] sender, have nowhere to put packet: {:?}", port, err);
+                    }
+                    if let Ok(task) = self.wake_providers[port].try_recv() {
+                        task.notify();
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Split Element Provider, exactly the same as AsyncElementProvider, but
+/// they have different trait bounds. Hence the reimplementaton. Would love
+/// a PR that solves this problem.
+pub struct ClassifyElementProvider<E: ClassifyElement> {
+    from_consumer: crossbeam_channel::Receiver<Option<E::Packet>>,
+    await_consumer: crossbeam_channel::Sender<task::Task>,
+    wake_consumer: crossbeam_channel::Receiver<task::Task>,
+}
+
+impl<E: ClassifyElement> ClassifyElementProvider<E> {
+    fn new(
+        from_consumer: crossbeam_channel::Receiver<Option<E::Packet>>,
+        await_consumer: crossbeam_channel::Sender<task::Task>,
+        wake_consumer: crossbeam_channel::Receiver<task::Task>
+    ) -> Self {
+        ClassifyElementProvider {
+            from_consumer,
+            await_consumer,
+            wake_consumer
+        }
+    }
+}
+
+impl<E: ClassifyElement> Drop for ClassifyElementProvider<E> {
+    fn drop(&mut self) {
+        if let Ok(task) = self.wake_consumer.try_recv() {
+            task.notify();
+        }
+    }
+}
+
+impl<E: ClassifyElement> Stream for ClassifyElementProvider<E> {
+    type Item = E::Packet;
+    type Error = ();
+
+    ///Implement Poll for Stream for AsyncElementProvider
+    /// 
+    /// This function, tries to retrieve a packet off the `from_consumer`
+    /// channel, there are four cases: 
+    /// ###
+    /// #1 Ok(Some(Packet)): Got a packet. If the consumer needs, (likely due to 
+    /// an until-now full channel) to be awoken, wake them. Return the Async::Ready(Option(Packet))
+    /// 
+    /// #2 Ok(None): this means that the consumer is in tear-down, and we
+    /// will no longer be receivig packets. Return Async::Ready(None) to forward propagate teardown
+    /// 
+    /// #3 Err(TryRecvError::Empty): Packet queue is empty, await the consumer to awaken us with more
+    /// work, and return Async::NotReady to signal to runtime to sleep this task.
+    /// 
+    /// #4 Err(TryRecvError::Disconnected): Consumer is in teardown and has dropped its side of the
+    /// from_consumer channel; we will no longer receive packets. Return Async::Ready(None) to forward
+    /// propagate teardown.
+    /// ###
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.from_consumer.try_recv() {
+            Ok(Some(packet)) => {
+                if let Ok(task) = self.wake_consumer.try_recv() {
+                        task.notify();
+                }
+                Ok(Async::Ready(Some(packet)))
+            },
+            Ok(None) => {
```

Comment:
It's a shame we can't use `try_ready!` here since we need the custom logic for the `Err(TryRecvError::Empty)` arm.